pkg/controller/core.oam.dev/v1alpha2/application/apply.go

```diff
@@ -529,3 +490,196 @@ func (h *appHandler) removeResourceTracker(ctx context.Context) (bool, error) {
+// get name of crossNamespace trait. The logic of this func is same with the way of appConfig generating traitName
+func (h *appHandler) getTraitName(ctx context.Context, componentName string, ct *v1alpha2.ComponentTrait, t *runtime.RawExtension) (string, error) {
+	trait, err := oamutil.RawExtension2Unstructured(t)
+	if err != nil {
+		return "", err
+	}
+	traitDef, err := oamutil.FetchTraitDefinition(ctx, h.r, h.r.dm, trait)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return "", errors.Wrapf(err, "cannot find trait definition %q %q %q", trait.GetAPIVersion(), trait.GetKind(), trait.GetName())
+		}
+		traitDef = oamutil.GetDummyTraitDefinition(trait)
+	}
+	traitType := traitDef.Name
+	if strings.Contains(traitType, ".") {
+		traitType = strings.Split(traitType, ".")[0]
+	}
+	traitName := oamutil.GenTraitName(componentName, ct, traitType)
+	return traitName, nil
+}
+
+// recodeTrackedResource append cross namespace resource to apphandler's acrossNsResources field
+func (h *appHandler) recodeTrackedResource(resourceName string, resource runtime.RawExtension) error {
+	u, err := oamutil.RawExtension2Unstructured(&resource)
+	if err != nil {
+		return err
+	}
+	tr := new(v1beta1.TypedReference)
+	tr.Name = resourceName
+	tr.Namespace = u.GetNamespace()
+	tr.APIVersion = u.GetAPIVersion()
+	tr.Kind = u.GetKind()
+	h.acrossNsResources = append(h.acrossNsResources, *tr)
+	return nil
+}
+
+// Now if workloads or traits are in the same namespace with application, applicationContext will take over gc workloads and traits.
+// Here we cover the case in witch a cross namespace component or one of its cross namespace trait is removed from an application.
+func (h *appHandler) garbageCollection(ctx context.Context) error {
+	rt := new(v1beta1.ResourceTracker)
+	err := h.r.Get(ctx, ctypes.NamespacedName{Name: h.generateResourceTrackerName()}, rt)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// guarantee app status right
+			h.app.Status.ResourceTracker = nil
+			return nil
+		}
+		return err
+	}
+	applied := map[v1beta1.TypedReference]bool{}
+	if len(h.acrossNsResources) == 0 {
+		h.app.Status.ResourceTracker = nil
+		if err := h.r.Delete(ctx, rt); err != nil {
+			return err
+		}
+		return nil
+	}
+	for _, resource := range h.acrossNsResources {
+		applied[resource] = true
+	}
+	for _, ref := range rt.Status.TrackedResources {
+		if !applied[ref] {
+			resource := new(unstructured.Unstructured)
+			resource.SetAPIVersion(ref.APIVersion)
+			resource.SetKind(ref.Kind)
+			resource.SetNamespace(ref.Namespace)
+			resource.SetName(ref.Name)
+			err := h.r.Delete(ctx, resource)
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					continue
+				}
+				return err
+			}
+		}
+	}
+	// update resourceTracker status, recode applied across-namespace resources
+	rt.Status.TrackedResources = h.acrossNsResources
+	if err := h.r.Update(ctx, rt); err != nil {
```

Comment:
```
h.r.UpdateStatus
```

Reply:
the status should be an sub resource. We should add a test case to cover it if the tests not fail.